src/ImageCore.jl

```diff
@@ -161,6 +161,7 @@ function PaddedViews.filltype(::Type{FC}, ::Type{C}) where {FC<:Colorant, C<:Col
     T = eltype(C) === Any ? eltype(FC) : eltype(C)
     _filltype(FC, base_colorant_type(C){T})
 end
+PaddedViews.filltype(::Type{FC}, ::Type{C}) where {FC<:AbstractGray, C<:Number} = base_color_type(FC){C}
```

Comment:
```julia
julia> PaddedViews.filltype(Gray24, Float32)
ERROR: TypeError: in Type{...} expression, expected UnionAll, got Type{Gray24}
```
(I personally don't like such type piracy, but I'm not much concerned with making decisions within JuliaImages.)

Reply:
"Glue packages" are a recognized way of combining functionality. The alternative would be to move this to `PaddedViews` under a `@require ColorTypes="..."` block.

Reply:
In this case, I think it's better to change the fallback in PaddedViews to the one based on `promote_type`, and add promotion rules between grays and numbers in ColorTypes. I don't think the performance issue mentioned in the PaddedViews comments are really fatal now.

Reply:
Unlike `promote_type`, `filltype` isn't symmetric:

```julia
julia> filltype(RGB{Float32}, Lab{Float32})
Lab{Float32}

julia> filltype(Lab{Float32}, RGB{Float32})
RGB{Float32}
```

When users call `PaddedView(c, a, indices)`, the promise of `filltype` is to let `eltype(a)` dominate the result type whenever possible.

Reply:
I meant something like:
```julia
function filltype(::Type{FT}, ::Type{T}) where {FT, T}
    Tp = promote_type(FT, T)
    isconcretetype(Tp) ? Tp : T
end
```
I don't know the practical cases where a fatal size increase and stall occurs. At least, if the user needs speed, all they have to do is specify the type manually.

I think that `mosaicview` is primarily intended for display, i.e. it will not be executed repeatedly.

Reply:
I get your point, I've thought about `Tp = promote_type(FT, T)` thing when I initialized the `fill_type`. But given that `PaddedView` was firstly designed as

```julia
PaddedView(fillvalue, data::AbstractArray{T,N}, indices) where {T,N} =
    PaddedView{T,N,typeof(indices),typeof(data)}(convert(T, fillvalue), data, indices)
```

Using `promote_type` might improve the code reusability in some sense, but that doesn't count a need to break the compatibility.

---

Good to know the existence of `isconcretetype`.

Reply:
> I bet they are sensitive to the performance.

I also bet they are sensitive to the numerical errors. :stuck_out_tongue:

I made a suggestion to change the "fallback". You can add rules without type piracy.
```julia
filltype(::Type{FT}, ::Type{T}) where {FT<:Real, T<:Number} = T
```

Reply:
As noted by @timholy, type piracy in this case isn't an issue. Right now I prefer to keep this as it is, when there's a real need we can switch to the ruling-out manner.